### PR TITLE
Onboard Qwen3.6-27B for four-node Slime training

### PR DESCRIPTION
**Pull request by cskitty**
## What changed

- add `Qwen3_6_27B` with Qwen3.6 response parsing and hybrid Gated DeltaNet/attention architecture metadata
- add one canonical `Qwen3_6_27b_Recipe` for 4 nodes × 8 H100s
- register and export the model/recipe through the normal Training Gym surfaces and API reference
- add the generated `003_qwen27b` multi-node DAPO-Math tutorial, following the existing Kimi multi-node tutorial structure
- add focused tests for topology, rollout/training settings, conversion layout, default-recipe registration, and colocated prefill-graph behavior

## Recipe

- trainer: TP=4, PP=2, CP=4, DP=1, sequence parallel; final PP stage receives 30 of 64 layers
- rollout: 16 SGLang TP=2 engines, batch 8, 8 samples/prompt, global batch 64, 32,768-token response cap
- memory/speculation: `mem_fraction_static=0.75`, EAGLE with 3 steps / top-k 1 / 4 draft tokens, extra-buffer Mamba scheduling
- training: GRPO with DeepScaler math reward, symmetric clip 0.2, no KL loss, per-token loss, LR `1e-6`, 8,192 training tokens/GPU
- optimizer: CPU-offloaded precision-aware Adam; optimizer state is saved, matching upstream Qwen3.5-27B
- conversion/checkpointing: torch-dist TP=4/PP=2, save every 20 steps plus the final checkpoint
- evaluation: disabled in the production preset

The topology and training/rollout values follow Slime's maintained Qwen3.5-27B recipe while loading `Qwen/Qwen3.6-27B`. Training Gym supplies its normal Modal cluster launch, volume paths, checkpoint conversion, status hooks, and dataset fields.

## EAGLE / MTP

Qwen3.6-27B declares one MTP layer and publishes trained `mtp.*` weights. SGLang loads those embedded weights from the HF checkpoint as the EAGLE draft head; Megatron does not instantiate or serialize an MTP training module (`mtp_num_layers` remains unset), avoiding the distributed torch-dist MTP conversion path.

The 10-step run logged `Qwen3_5ForCausalLMMTP` loads of about 2.82 GB across rollout engines and active speculative decoding with approximately 3.1–3.5 accepted tokens per verification pass and engine acceptance rates around 70–84%.

### 10-step end-to-end run

- run: [`candied-fleece-9cf7e02ff5d5`](https://modal-labs-melody-dev--training-gym-dashboard-fastapi-app.modal.run/training/candied-fleece-9cf7e02ff5d5)
- hardware: 4 nodes × 8 H100s
- completed all 10 rollout, log-prob, optimizer, weight-sync, and final-checkpoint phases
- 64 scored samples per step; rewards: `0.594, 0.797, 0.859, 0.484, 0.828, 0.719, 0.688, 0.531, 0.766, 0.828` (mean `0.709`)
- rollout generation averaged 172.3 seconds/step (range 158.9–202.9 seconds)
- no OOMs, NaNs, malformed responses, or distributed failures; inspected responses remained coherent and task-directed
- the end-to-end run included first-time HF→torch-dist conversion and the first optimizer's cold compilation; subsequent optimizer phases settled to the expected steady-state range

The run used `num_rollout=10` as a validation-only override. It was launched before evaluation was disabled, so it also performed an initial smoke evaluation; the checked-in tutorial uses the unmodified production `Qwen3_6_27b_Recipe()`.

Additional validation:

- separate 4-step topology/run-path canary completed without OOMs or malformed output
- separate 8-step learning canary completed all updates and improved its held-out eval from 40/64 (`0.625`) to 46/64 (`0.71875`)
- focused model-config tests: 2 passed
- Ruff lint and format checks passed
- generated tutorial Python compiles, regeneration is deterministic, and `git diff --check` passes

## Scope

This validates text-only causal-language-model RL. Qwen3.6's vision encoder is represented in the upstream checkpoint but multimodal batching/training is not exercised by this recipe or tutorial.

### CODE REVIEW

File: modal_training_gym/common/models/qwen3_6_27b.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        kv_channels=256,
+        vocab_size=248320,
+        normalization="RMSNorm",
+        norm_epsilon=1e-6,
+        swiglu=True,
+        disable_bias_linear=True,
+        qk_layernorm=True,
+        untie_embeddings_and_output_weights=True,
+        megatron_spec=["slime_plugins.models.qwen3_5", "get_qwen3_5_spec"],
+        megatron_model_type="qwen3.5-27B",
```
[the remark this answers is not in the document]
Comment (cskitty) [reply modal_training_gym/common/models/qwen3_6_27b.py:33]: This is deltanet/attention not standard MHA

**Review by joyliu-q — APPROVED** [review]
